### ISSUE TITLE
Unable to use paths with spaces in PathMap

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#22835]
Version used: 15.5. Preview 2 (27019.1.d15rel)

Repro:

Create a new .Core console app in a directory that contains spaces (e.g. ```C:\Users\{user}\Documents\Visual Studio 2017\Projects\ConsoleApp1```):

```C#
using System;
using System.Runtime.CompilerServices;

class Program
{
    static string GetPath([CallerFilePath]string s = null) => s;
    static void Main(string[] args) => Console.WriteLine(GetPath());
}
```

Edit the project file and add:

```xml
  <PropertyGroup>
    <PathMap>$(MSBuildProjectDirectory)=Q:\a b</PathMap>
  </PropertyGroup>
```

Build and run the app. It should print: 

```
Q:\a b\Program.cs
```

but prints 

```
C:\Users\{user}\Documents\Visual Studio 2017\Projects\ConsoleApp1\ConsoleApp1\Program.cs
```

which means the PathMap wasn't applied.

Setting the value to ```<PathMap>"$(MSBuildProjectDirectory)=Q:\a b"</PathMap>``` or  ```<PathMap>"$(MSBuildProjectDirectory)"="Q:\a b"</PathMap>``` does not work either.

Setting the value to ```<PathMap>C:\=Q:\</PathMap>``` works as expected.

**Comment by tmat** [comment]
@jaredpar @gafter

**Comment by jcouv** [comment]
@khyperia has a PR that allows spaces in pathMaps from command-line. @tmat can you take a look and confirm whether it addresses your screnario from msbuild too?
https://github.com/dotnet/roslyn/pull/23053

**Comment by tmat** [comment]
@jcouv @khyperia The problem with spaces is fixed but there is still problem of quoting. 

It is currently not possible to map path that contains ```,``` or ```=``` from msbuild. 

In the above example if I use quotes around the path they get escaped like so:
```<PathMap>"$(MSBuildProjectDirectory)"="Q:\a,=b"</PathMap>```

csc command line:

```
/pathmap:"\"c:\users\tomat\documents\visual studio 2017\Projects\ConsoleApp103\ConsoleApp103\"=\"Q:\a,=b\""
```

Which breaks the mapping.

**Comment by paul1956** [comment]
Not sure if this is where to post this but Roslyn's Restore.cmd also is broken in 15.5 with space in path to roslyn-master. This is a regression.

**Comment by jaredpar** [comment]
@paul1956 

> Not sure if this is where to post this but Roslyn's Restore.cmd also is broken in 15.5 with space in path to roslyn-master. This is a regression.

It's possible this worked before but it's something we explicitly don't support at this time. Here is the full list of restrictions we have for the enlistment path in Roslyn. 

https://github.com/dotnet/roslyn/issues/20929

**Comment by tmat** [comment]
@jaredpar Can we get this fixed in 15.6? I'm working on integrating deterministic path map into Common.targets and would like to adding checks that error in presence of ```,``` and ```=``` in the paths.

**Comment by gafter** [comment]
See #22835 for additional details and repro.

See also #24247 for a related proposed change.

**Comment by tmat** [comment]
@jaredpar @gafter Can we get this fixed in 16.7? We are driving deterministic build adoption for packages in the .NET ecosystem and this is one of the issues that people might run into. We want to make it as seamless as possible for package authors to switch to fully deterministic builds and bugs like these are getting in the way.

/cc @clairernovotny @richlander
